Fixed an issue with shorter param list not being assignable to rest param with tuple union

Andarist opened the pull request:
fixes https://github.com/microsoft/TypeScript/issues/48663

Andarist commented on src/compiler/checker.ts:
```diff
@@ -17967,7 +17967,10 @@ namespace ts {
                 const sourceType = i === restIndex ? getRestTypeAtPosition(source, i) : tryGetTypeAtPosition(source, i);
-                const targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+                let targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+                if (i === restIndex && targetType && sourceType && isTupleType(sourceType) && !sourceType.target.hasRestElement) {
```
I have concluded that `sourceType.target.hasRestElement` doesn't have to come with any special handling here but perhaps there is a case that could be handled here - if you come up with any I might try to accommodate it in the logic.

Andarist commented on src/compiler/checker.ts:
```diff
@@ -17967,7 +17967,10 @@ namespace ts {
-                const targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+                let targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+                if (i === restIndex && targetType && sourceType && isTupleType(sourceType) && !sourceType.target.hasRestElement) {
+                    targetType = mapType(targetType, t => isTupleType(t) && !t.target.hasRestElement ? sliceTupleType(t, 0, t.target.fixedLength - sourceType.target.fixedLength) : t);
```
Ideally, I should drop `!t.target.hasRestElement` from here because I think that rest in the target tuple could be handled. For instance this should typecheck OK:
```ts
// Type '(a: string | number, b: string | number) => void' is not assignable to type '(...args: [number, string, ...boolean[]] | [string, number, ...boolean[]]) => void'.
//   Types of parameters 'a' and 'args' are incompatible.
//     Type '[number, string, ...boolean[]] | [string, number, ...boolean[]]' is not assignable to type '[a: string | number, b: string | number]'.
//       Type '[number, string, ...boolean[]]' is not assignable to type '[a: string | number, b: string | number]'.
//         Target allows only 2 element(s) but source may have more.(2322)
const f: (
  ...args: [number, string, ...boolean[]] | [string, number, ...boolean[]]
) => void = (a, b) => {};
```

This would require a different kind of a tuple slice - or rather, a tuple expansion (?). I think that the correct answer would be to convert a tuple type with rest to a tuple type of completely fixed elements (since we are trying to match a fixed-size source). 

I've looked throughout the codebase for an algorithm like that and I couldn't find it - perhaps you would know if something like this is already implemented anywhere? If this isn't implemented anywhere then I can probably write it, I've just figured out that it might be best to first ask about it cause it feels that it might already exist somewhere.

jakebailey replied:
It feels like this should exist, given one can write:

```ts
type AtLeastOne = [number, ...number[]];
type AtLeastTwo = [number, number, ...number[]];
type AtLeastThree = [number, number, number, ...number[]];

declare let one: AtLeastOne;
declare let two: AtLeastTwo;
declare let three: AtLeastThree;
declare let twoOrThree: AtLeastTwo | AtLeastThree;

one = two;
one = three;
one = twoOrThree;
```

In that the function can accept fewer parameters and it's legal, just like assigning a "bigger" tuple is to a smaller tuple that optionally takes more.

The "Target allows only 2 element(s) but source may have more.(2322)" message is the bug that this PR is trying to fix, but the fact that it only works in one way (if I am understanding this correctly) leaves me a little suspicious that there might be something missing.

Andarist replied:
Thanks for the given snippet! I will analyze this and compare how the compiler checks those. I'm a little bit busy right now but I hope to get back to this PR within 2 weeks.

Andarist replied:
I'm slowly looking back at this since I'm also trying to fix a different, related, issue. In your examples, all those types have a variadic length whereas all (or most) of the situations that I already handle here and which I was thinking about here include a fixed size tuple,

Weirdly enough I have just discovered that this is not a valid assignment today:
```ts
declare let atLeastOne: [number, ...number[]];

const justOne: [number] = atLeastOne
// Type '[number, ...number[]]' is not assignable to type '[number]'.(2322)
```

I don't understand why this limitation is in place here as intuitively I would consider a type with *more* keys to be always assignable to one with *less* keys. After all, this works perfectly fine:
```ts
const foo: { a: 1 } = {} as { a: 1, b: 2 }
```

Andarist replied:
I guess the reason why we can't assign to fixed size tuples like this is their `length` property that is typed as a discrete number.

graphemecluster commented:
I've been working on a library that's been plagued by #48663, so I'm really happy to see this PR.
And will #45972 be fixed by this PR as well?

Andarist commented:
> And will https://github.com/microsoft/TypeScript/issues/45972 be fixed by this PR as well?

Not yet - but I think it has the potential to do that. I intentionally didn't handle this case yet because I wasn't sure how to properly "expand" the target type to a fixed-size tuple. I've mentioned this in the comment [here](https://github.com/microsoft/TypeScript/pull/49218#discussion_r879757992)

Andarist commented:
Actually, my simplistic "expand" idea wouldn't work with leading/middle rest elements. However, one could write a correct logic based on the code here:
https://github.dev/microsoft/TypeScript/blob/18193973fec66716b1489d3edeeb4a3c837bd39f/src/compiler/checker.ts#L20128-L20198

It is a little bit convoluted though, gonna continue looking into this.

graphemecluster commented:
I am actually not familiar with the checker, but I wonder if this PR could also expand all the tuple arguments to its function overload forms (currently it only works if it is not a union type).
For example,

```ts
(...args: [string] | [string, number]) => void
			↓
(args_0: string, args_1?: number) => void

(...args: [string, number] | [number, string]) => void
			↓
(1/2) (args_0: string, args_1: number) => void
(2/2) (args_0: number, args_1: string) => void

(...args: [string, number, ...boolean[]] | [number, string, ...boolean[]]) => void
			↓
(1/2) (args_0: string, args_1: number, ...args_2: boolean[]) => void
(2/2) (args_0: number, args_1: string, ...args_2: boolean[]) => void

(...args: [...boolean[], string, number] | [...boolean[], number, string]) => void
			↓
(1/2) (...args_0: boolean[], args_1: string, args_2: number) => void
(2/2) (...args_0: boolean[], args_1: number, args_2: string) => void
```

For the last case, I still find it easier to read although it is not valid JavaScript.

Andarist commented:
That is definitely out of the scope of this PR. I don't intend to work on something like this - also, once you start involving generics in this then unions and overloads might stop being the same. Union means one of the members - any member, it's order-independent. OTOH, overloads are always matched from the top to the bottom so the first match "wins".

graphemecluster commented:
In what cases do matching the tuples not imply the match for the corresponding function overloads? And is it worth opening an issue for this (if it won’t be a duplicate)? Sorry to keep asking you questions.

graphemecluster commented:
I think the easiest way to solve all the related issues is to wrap all the arguments into a tuple with rest syntax, i.e. `item => {}` should become `(...[item]) => {}` and `([first, ...rest], foo, ...args) => {}` should work just like `(...[[first, ...rest], foo, ...args]) => {}`.

Andarist commented on tests/baselines/reference/genericRestParameters3.errors.txt:
```diff
@@ -69,16 +70,17 @@ tests/cases/conformance/types/rest/genericRestParameters3.ts(59,5): error TS2345
-!!! error TS2322:     Type '[string] | [number, boolean]' is not assignable to type '[y: number, z: boolean]'.
-!!! error TS2322:       Type '[string]' is not assignable to type '[y: number, z: boolean]'.
-!!! error TS2322:         Source has 1 element(s) but target requires 2.
+!!! error TS2322:     Type '[number, boolean] | [string, undefined]' is not assignable to type '[y: number, z: boolean]'.
+!!! error TS2322:       Type '[string, undefined]' is not assignable to type '[y: number, z: boolean]'.
+!!! error TS2322:         Type at position 0 in source is not compatible with type at position 0 in target.
+!!! error TS2322:           Type 'string' is not assignable to type 'number'.
```
This error change comes from the fact that I fill the slice of the target type with undefineds here (when it's shorter than the source's slice):
https://github.dev/microsoft/TypeScript/blob/6c5383d701417e8a98398b4171af0f201ee593c5/src/compiler/checker.ts#L19394-L19402

I think this is a good change. This will allow more patterns to be assignable, especially with target signatures using unions of tuples for their rest parameter. After all, the implementation is free to ignore any of the provided arguments etc. This is also already just fine in TS:
```ts
declare let f1: (a: string) => void
declare let f2: (a: string, b: number) => void

f2 = f1
```

And since this is OK I think that this one should be too:
```ts
declare let f1: (x: string, ...args: [string] | [number, boolean]) => void;

// Type '(a: string, b: string | number, c: boolean | undefined) => void' is not assignable to type '(x: string, ...args: [string] | [number, boolean]) => void'.
//   Types of parameters 'b' and 'args' are incompatible.
//     Type '[string] | [number, boolean]' is not assignable to type '[b: string | number, c: boolean | undefined]'.
//       Type '[string]' is not assignable to type '[b: string | number, c: boolean | undefined]'.
//         Source has 1 element(s) but target requires 2.
f1 = (a, b, c) => {}
```

So with this change that caused this particular error here to be reported differently we now allow the example above and I added a test case for it:
https://github.com/microsoft/TypeScript/blob/46f2e57d76f26b74f596f2acb5d04a30572f13a7/tests/cases/compiler/restTupleUnionWithRestContextualParams.ts#L6

Andarist commented on tests/baselines/reference/genericRestParameters3.errors.txt:
```diff
@@ -69,16 +70,17 @@ tests/cases/conformance/types/rest/genericRestParameters3.ts(59,5): error TS2345
-!!! error TS2322:     Type '[string] | [number, boolean]' is not assignable to type '[y: string]'.
-!!! error TS2322:       Type '[number, boolean]' is not assignable to type '[y: string]'.
-!!! error TS2322:         Source has 2 element(s) but target allows only 1.
+!!! error TS2322:     Type '[string] | [number]' is not assignable to type '[y: string]'.
+!!! error TS2322:       Type '[number]' is not assignable to type '[y: string]'.
+!!! error TS2322:         Type 'number' is not assignable to type 'string'.
```
This is a change of the error for this case:
```ts
declare let f1: (x: string, ...args: [string] | [number, boolean]) => void;
declare let f2: (x: string, y: string) => void;

f1 = f2
```

With the current changes in this PR, I create a slice from the target tuple and I cap it to the source's length~. I do that for similar reasons that I've outlined in the other comment. The source signature can freely ignore the "extra" arguments so we don't even need to check them here (we'd have to ignore them through some other mechanism anyway)

Andarist commented:
@jcalz this PR implements changes for the problem that you have noticed [here](https://github.com/microsoft/TypeScript/pull/47190#issuecomment-1339753554). I pushed out tests for those cases in [this commit](https://github.com/microsoft/TypeScript/pull/49218/commits/0de8923abaa3424e113cc58f33c5b9480a2d9f1e)

Andarist commented:
@jakebailey my "I'll get back to this within 2 weeks" took a little bit longer 🤣 but I've done some work over the holidays here and I think this PR is in pretty good shape right now. Would you mind taking another look at this?

gabritto commented on src/compiler/checker.ts:
```diff
@@ -19381,7 +19381,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const sourceType = i === restIndex ? getRestTypeAtPosition(source, i) : tryGetTypeAtPosition(source, i);
-            const targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            let targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            if (i === restIndex && targetType && sourceType && isTupleType(sourceType)) {
```
You're intentionally not fixing the case where `sourceType` is a union, right?

Andarist replied:
> You're intentionally not fixing the case where sourceType is a union, right? I think that's ok because I think the code would become a lot more complicated, but I wanted to ask.

Yeeah. It's likely that this could be fixed by wrapping the body of this loop, executing it for each union member, and checking if any of the iterations return ~ `Ternary.False`. Maybe that could be explored in a follow-up PR - but for the time being, I think that it's best to improve this incrementally. It's way easier to follow this diff here than it would be to follow it if I would try to handle union sources as well.

> Same for when targetType is an array I guess.

Hmm, I think this is actually already covered - it's just that for this case we don't need to do anything special, so we forward the type so it can be checked by the latter logic as-is. Maybe I'm missing something though - do you have any particular test case in mind? The main purpose behind the added code is to "adjust" the target tuple types - their lengths etc, as without that we simply fail on required elements checks etc (but when comparing signatures we can actually often ignore extra positions that are not utilized by the source signature).

gabritto commented:
@typescript-bot pack this

gabritto commented on src/compiler/checker.ts:
```diff
@@ -19381,7 +19381,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const sourceType = i === restIndex ? getRestTypeAtPosition(source, i) : tryGetTypeAtPosition(source, i);
-            const targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            let targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            if (i === restIndex && targetType && sourceType && isTupleType(sourceType)) {
```
I think I have a reasonable understanding now, so feel free to ignore the question, but it might still be helpful to have a sort of high level description.

Andarist replied:
I added source comments in which I try to explain what is happening here, let me know what do you thnk about them.

gabritto reviewed (changes requested):
In general, the algorithm we use for signature comparability is currently not complete, and I think not intended to be complete, leading to the bug reports cited in this PR. The proposed bug fix in the PR is also not complete, but I think that's not an issue as long as the PR can solve the problem in some cases. As demonstrated by the new tests, it does fix some of the weird cases we currently have. But I also tried to think about two things:

(1) Does it break existing code?
I found that the following code errors with this PR, but works fine on main:
`const exh: <T extends string[]>(...rest: [number, ...T, boolean] ) => void = (a: number, ...arg: [...string[], boolean]) => {};`
[Playground link](https://www.staging-typescript.org/play?ts=5.0.0-pr-49218-12#code/MYewdgzgLgBApgDwBYC4YB4Aq8FTmAEwhmgCcBLMAcwG0BdAPgAoA6N0uaNGsAVwFsARnFIAaGGxaZxgkCAA2cAIZg6MAJQwAvAxgA3EOQLaYTJWj5CR4yUtJVukspVp0ZcxSrqadMAN4AvgDcQA)
I haven't checked if this is a bug with the new code, or an inherent limitation of the new approach.

(2) Is it worth the maintenance cost?
I don't know, it's not a lot of new code, but we still could look at this code in the future and ask ourselves why we need it. Maybe a comment could help with that.

Andarist commented:
> I haven't checked if this is a bug with the new code, or an inherent limitation of the new approach.

Looks like a bug to me - thanks for noticing this! I'm heading to a company retreat just now, so I will work on fixing this next week.

> I don't know, it's not a lot of new code, but we still could look at this code in the future and ask ourselves why we need it. Maybe a comment could help with that.

I'm happy to add some comments about this to the source code.

Andarist commented:
@gabritto I fixed the case that you spotted during the review and added a few other tests as well (similar ones to what you observed). 

I also added some source comments - let me know if that helps to understand the purpose of this code.

> (2) Is it worth the maintenance cost?

I was also thinking about this since it's a totally valid concern. I think that there this PR has a pretty good RoI. It fixes some unintuitive behavior, the introduced logic stays fairly self-contained (just one extra "processing" step added), and the introduced logic is actually not that complicated (when we consider the overall complexity of the compiler). So I think that this cost is justified here. For comparison - when I was working on this PR https://github.com/microsoft/TypeScript/pull/51157 (that was also related to tuples), it took me way more time to understand the existing code and the adjustments that I had to introduce there than it took me to get this one here going.

Andarist commented on src/compiler/checker.ts:
```diff
@@ -19381,7 +19381,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+
+                    for (let i = 0; i < getTypeReferenceArity(sourceType); i++) {
+                        elementTypes.push(getTupleElementType(t, i)!);
+                        elementFlags.push(sourceType.target.elementFlags[i]);
```
That's an excellent question. At first, I wasn't sure if this is correct - it has been a while since I worked on this. I rechecked that and this is intentional. The implied source tuple type, which represents parameters of the source signature, might only have fixed elements - while the target tuple type might be declared using rest. For example:
```ts
const f1: (...args: [number, string, ...boolean[]] | [string, number, ...boolean[]]) => void = (a, b, c) => {};
```

Here, we want to compare those *contravariant* tuples:
```ts
type Target = [number, string, boolean] | [string, number, boolean]
type Source = [a: string | number, b: string | number, c: boolean]
```

However, at the moment - I'm not entirely sure if all "transitions" that can happen here are OK. I need to analyze this again and think if it's always OK to, for example, create an optional element from some required elements.

gabritto replied:
I'm trying to review this again, and I'm trying to think if using the source tuple type's element flags to construct the new target tuple types is ok. I was trying to think through the case where the source element is optional, so the new target element will be too. I think this makes sense and is OK to do when we're comparing function signatures strictly, i.e. when we'll compare parameter types contravariantly. However, if we're comparing *method* signatures, we'll end up comparing the parameter types also covariantly, and in that case I don't know if it makes sense for the new target type element to be marked as optional for that comparison.

A perhaps convoluted example I came up with to try and show this:
```ts
// Counter-example to making target element optional when source element is optional?
type NowSource<T extends boolean[]> = [...rest: [boolean, ...T, n?: number]];
type NowTarget<T extends boolean[]> = [...rest: [boolean, ...T, number | string]];

class A {
  method<T extends boolean[]>(...args: NowTarget<T>): void {}
}

class B extends A {
  method<T extends boolean[]>(...args: NowSource<T>): void {} // Errors on main, no error on PR
}

declare let func: <T extends boolean[]>(...args: NowTarget<T>) => void;
declare let gunc: <T extends boolean[]>(...args: NowSource<T>) => void;

func = gunc; // Errors on main and PR, ok
```

Andarist commented on src/compiler/checker.ts:
```diff
@@ -19381,7 +19381,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                    const elementFlags: ElementFlags[] = [];
+
+                    for (let i = 0; i < getTypeReferenceArity(sourceType); i++) {
+                        elementTypes.push(getTupleElementType(t, i)!);
```
It's because `getTupleElementType` guarantees a `Type` to be returned for `TupleTypeReference` arguments:
https://github.dev/microsoft/TypeScript/blob/4ba756fdc847115bc1ff5032ed06e9bed60f16b2/src/compiler/checker.ts#L22757-L22759

Andarist commented on src/compiler/checker.ts:
```diff
@@ -22881,9 +22923,15 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
-    function isTupleTypeStructureMatching(t1: TupleTypeReference, t2: TupleTypeReference) {
+    function isTupleTypeStructureMatching(t1: TupleTypeReference, t2: TupleTypeReference, strictVariableElementsComparison: boolean) {
```
I'm not super fond of this change. It feels a little bit icky that we end up with different "modes" for comparing tuple type structures.

However, the name of the function is almost perfect for my use case - I couldn't figure out a better name for a "forked" implementation of this one. There is only one existing call to this helper function and it can be found here:
https://github.dev/microsoft/TypeScript/blob/4ba756fdc847115bc1ff5032ed06e9bed60f16b2/src/compiler/checker.ts#L24408-L24415

Note that this whole function was already a little bit specific to this one call site - since that one call site **allows** a required element to match an optional one. So, in a sense, I do a similar thing by allowing a variadic one to match a rest one. 

I'm also debating with myself - if in my case, I shouldn't disallow required elements from matching optional ones. I need to think about this more though, perhaps I will find a different solution for the cases related to the optional elements that I have to fix. If it comes to that - then I will probably refactor this new parameter to accept a bitmask.

Andarist commented on tests/baselines/reference/dependentDestructuredVariables.errors.txt:
```diff
@@ -1,8 +1,13 @@
 tests/cases/conformance/controlFlow/dependentDestructuredVariables.ts(334,5): error TS7022: 'value1' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
 tests/cases/conformance/controlFlow/dependentDestructuredVariables.ts(334,5): error TS7031: Binding element 'value1' implicitly has an 'any' type.
+tests/cases/conformance/controlFlow/dependentDestructuredVariables.ts(408,7): error TS2322: Type '(type: "one" | "two", x: string | number | undefined) => void' is not assignable to type '(...args: [type: "one", x?: number | undefined] | [type: "two", x: string]) => void'.
```
This is a case that regressed with [the newest commit](https://github.com/microsoft/TypeScript/pull/49218/commits/e3574503013a5dc983967a7ea30841ce4be3f7d9). I'm still thinking how to properly fix this - perhaps the solution is to disallow optional and required elements matching (as I mentioned in [this comment](https://github.com/microsoft/TypeScript/pull/49218/files#r1071104294)). It's a little bit fuzzy to me how optional elements followed but variable ones should behave, so I need to find a moment to write down some test cases and try out the idea with those in place.

Andarist commented on tests/baselines/reference/restTupleUnionWithRestContextualParams.errors.txt:
```diff
@@ -0,0 +1,43 @@
+tests/cases/compiler/restTupleUnionWithRestContextualParams.ts(7,7): error TS2322: Type '(type: "one" | "two", x: string | number | undefined) => void' is not assignable to type '(...args: [type: "one", x?: number | undefined] | [type: "two", x: string]) => void'.
```
This is essentially the same regression that I mentioned [here](https://github.com/microsoft/TypeScript/pull/49218/files#r1071106463). Note that this is only a regression when I take into account the previous version of this PR - this test case doesn't typecheck on the main branch.

ssalbdivad commented:
+1 to assignability improvements in this area.

Was struggling with these errors today on some very simple spread tuple unions:

```ts
const f = (...args: [unknown]) => args

declare const input: [unknown] | [unknown, unknown]

// A spread argument must either have a tuple type or be passed to a rest parameter.ts(2556)
f(...input)
```

Unfortunately, you also can't very easily cast your way out of this, as trying to spread `any` or `never` or some non-fixed length array results in the same message.

Not sure if this is already addressed by the changes here, but it would be really nice if in simple cases like this TS were able to determine the args were safe, or at least if there was some easier workaround.

Andarist commented:
@ssalbdivad the presented case isn't being fixed here. I think this one stems from the fact that this isn't valid:
```ts
const f = (arg: unknown) => {}
declare const input: [unknown, unknown]
// Expected 1 arguments, but got 2.(2554)
f(...input)
```

The surplus of arguments isn't allowed at the call sites today.

jakebailey commented:
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster
@typescript-bot pack this

gabritto commented:
@typescript-bot run DT
@typescript-bot user test this
@typescript-bot test top100
@typescript-bot perf test this

gabritto commented:
@typescript-bot perf test this
@typescript-bot perf test this public

gabritto commented:
@typescript-bot pack this

jakebailey commented:
@typescript-bot test it

jakebailey commented on src/compiler/checker.ts:
```diff
@@ -20732,7 +20738,59 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                    //   b; // number | string
+                    // }
+                    // fn('str'); // valid
+                    const elementTypes: Type[] = [];
+                    const elementFlags: ElementFlags[] = [];
```
Typically we use `append` rather than pushing onto an empty list directly; this is typically more efficient due to how the runtime constructs arrays.

jakebailey commented on src/compiler/checker.ts:
```diff
@@ -24440,9 +24498,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+            every(t1.target.elementFlags, (f1, i) => {
+                const f2 = t2.target.elementFlags[i];
+                return f1 === f2 ||
+                    !!(tupleStructureComparisonKind & TupleStructureComparisonKind.MatchFixed && f1 & ElementFlags.Fixed && f2 & ElementFlags.Fixed) ||
+                    !!(tupleStructureComparisonKind & TupleStructureComparisonKind.MatchVariable && f1 & ElementFlags.Variable && f2 & ElementFlags.Variable);
```
Why aren't these something like `f1 & ElementFlags.Fixed === f2 & ElementFlags.Fixed`, as the original code did?

jakebailey commented on src/compiler/checker.ts:
```diff
@@ -24440,9 +24498,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
-    function isTupleTypeStructureMatching(t1: TupleTypeReference, t2: TupleTypeReference) {
+    function isTupleTypeStructureMatching(t1: TupleTypeReference, t2: TupleTypeReference, tupleStructureComparisonKind: TupleStructureComparisonKind) {
```
Curious if we could just accept `ElementFlags` as a mask here instead of coming up with a new enum.

gabritto commented:
@typescript-bot pack this

gabritto reviewed (changes requested):
Ok I did review this a second time earlier this year, and I left a comment with a case that I think might be wrong with this PR, but forgot to "request changes": https://github.com/microsoft/TypeScript/pull/49218#discussion_r1446887889

Sorry about that.

sandersn commented:
This has been sitting for a while with changes requested. @Andarist do you want to keep working on it?
